### PR TITLE
Reprovision when identity reconciliation fails

### PR DESCRIPTION
**Pull request by gordonwang0**
If Identity Service fails to contact Hub during identity reconciliation, it will try once to reprovision the device.

### CODE REVIEW

File: identity/aziot-identityd/src/lib.rs [hunk 1]
```diff
@@ -412,10 +412,37 @@ impl Api {
...
+                    ReprovisionTrigger::Startup | ReprovisionTrigger::ConfigurationFileUpdate => {
+                        log::info!("Could not reconcile Identities with current device data. Reprovisioning.");
+
+                        self.id_manager
+                            .provision_device(self.settings.provisioning.clone(), false)
+                            .await?;
+                        log::info!("Successfully reprovisioned.");
+
+                        self.id_manager
+                            .reconcile_hub_identities(self.settings.clone())
```
Comment (massand) [inline identity/aziot-identityd/src/lib.rs:432]: Tracing is beneficial for the second reconcile attempt as well

Comment (gordonwang0) [reply identity/aziot-identityd/src/lib.rs:432]: If this fails, the error gets logged by the caller (and if it succeeds, the "Identity reconciliation complete message" below gets logged).

File: identity/aziot-identityd/src/identity.rs [hunk 2]
```diff
@@ -516,7 +516,11 @@ impl IdentityManager {
                             self.get_backup_provisioning_info(credential.clone())?;
 
                         if skip_if_backup_is_valid && backup_device.is_some() {
-                            backup_device.expect("backup device cannot be none")
+                            let backup_device =
```
Comment (massand) [inline identity/aziot-identityd/src/identity.rs:519]: Can we DRY these statements into the `operation_to_iot_hub_device` method and rename the method?

Comment (gordonwang0) [reply identity/aziot-identityd/src/identity.rs:519]: This log message is only printed when a valid backup exists; i.e. when IS is not making an operation to IoT Hub.

Comment (massand) [reply identity/aziot-identityd/src/identity.rs:519]: I forgot to tag my comment for all the statements from lines 515-527. I meant DRY all those lines for all DPS auth types. I think that can be done?

Comment (gordonwang0) [reply identity/aziot-identityd/src/identity.rs]: Got it. Did that.

**Review by massand — APPROVED** [review]

**Pull request merged**